File: Breadth-First Search/Topological sort/LET_no310_Minimum Height Trees.cpp
class Solution {
public:
    vector<int> findMinHeightTrees(int n, vector<vector<int>>& edges) {
        if (n == 1)
            return {0};
        
        vector<int> degree(n, 0);
        vector<vector<int>> adj(n);
        for (auto& e : edges) {
            adj[e[0]].push_back(e[1]);
            adj[e[1]].push_back(e[0]);
            degree[e[0]]++;
            degree[e[1]]++;
        }
        
        queue<int> q;
        for (int i = 0; i < n; i++) {
            if (degree[i] == 1)
                q.push(i);
        }
        
        vector<int> res;
        while (!q.empty()) {
            res.clear();
            int size = q.size();
            for (int i = 0; i < size; i++) {
                int cur = q.front();
                q.pop();
                res.push_back(cur);
                for (int& neighbor : adj[cur]) {
                    degree[neighbor]--;
                    if (degree[neighbor] == 1)
                        q.push(neighbor);
                }
            }
        }
        
        return res;
    }
};

File: Breadth-First Search/Topological sort/LET_no310_Minimum Height Trees.py
class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n == 1:
            return [0]

        degree = [0 for _ in range(n)]
        neighbors = collections.defaultdict(list)
        for a, b in edges:
            neighbors[a].append(b)
            neighbors[b].append(a)
            degree[a] += 1
            degree[b] += 1

        queue = collections.deque([i for i in range(n) if degree[i] == 1])
        res = None
        while queue:
            temp = []
            for _ in range(len(queue)):
                cur = queue.popleft()
                temp.append(cur)
                for neighbor in neighbors[cur]:
                    degree[neighbor] -= 1
                    if degree[neighbor] == 1:
                        queue.append(neighbor)
            res = temp
        
        return res
